### ISSUE TITLE
CGI plugin - Fails to recognise \r\n\r\n if split up over multiple reads

### ISSUE THREAD

**Issue by eccoecco** [monkey/monkey#202]
Monkey 1.5.6's CGI plugin appears to have trouble if the \r\n\r\n sequence is split over multiple reads.  For example:

    #!/bin/sh
    echo -e -n "Content-Type: text/plain\r\n"

    # Represents some other time consuming task - may take as little as 10-20ms, i.e. a context switch
    sleep 1s

    echo -e -n "\r\n"

    # To give a separate read event
    sleep 1s
    echo "Hello world"

**Expected behaviour:**
The CGI plugin finds the \r\n\r\n line break separating the header and body, so that it can insert the chunk size prior to the body.

**Actual behaviour:**
The CGI plugin does not find the \r\n\r\n, and "Hello World" is sent without a chunk size, causing a browser to complain about invalid chunk encodings.

**Suspected Cause:**
In plugins/cgi/event.c - _mkp_event_write():

        if (!r->all_headers_done)
        {
            unsigned char advance = 4;

            // Write the rest of the headers without chunking
            char *end = getearliestbreak(outptr, r->in_len, &advance);

The first read event fires, and causes getearliestbreak() to search for \r\n\r\n (or just \n\n) over the input data "Content-Type: text/plain\r\n".  It does not find the \r\n\r\n sequence.

The second read event fires, and cannot match the earliest break in the "\r\n" input data.

Finally, the third read even fires, and because the header hasn't been found, the CGI plugin treats "Hello world" as part of the header, instead of inserting a chunk size, and sending it out as a chunk.

**Possible Fix:**
I tried to fix it on my installation by doing this, but unfortunately unaware if there are any negative side effects:

            char *end = getearliestbreak(outptr, r->in_len, &advance);
            if (!end)
            {
                // Just in case the last few characters of the header is CRLFCR, and the LF is coming in the next read
                if (r->in_len > 3)
                {
                    swrite(socket, outptr, r->in_len - 3);
                    memmove(r->in_buf, outptr + r->in_len - 3, 3);
                    r->in_len = 3;
                }

                mk_api->event_socket_change_mode(socket, MK_EPOLL_SLEEP, MK_EPOLL_LEVEL_TRIGGERED);
                return MK_PLUGIN_RET_EVENT_OWNED;
            }

This just keeps the last 3 characters of the current buffer present, just in case those last three characters form part of the \r\n\r\n pattern with the next read.

As I have not done a full review of the consequences of this change, I leave it here as a possible solution, but acknowledge that it might be a naive solution.

**Comment by edsiper** [comment]
I am investigating the issue.

Also I am evaluating if we should or not continue supporting a CGI interface.

**Comment by ambiHomeCE** [comment]
I just ran into this issue myself. Is there a usable old version, which doesn't suffer from this problem? I'd love to see this fix and a continuation of the CGI interface.

**Comment by edsiper** [comment]
I'd be fine including a fix in our new major release Monkey v1.6. 

@ambiHomeCE Would you please provide more details of the context where do you use Monkey ?

**Comment by ambiHomeCE** [comment]
We use Monkey as web server on a raspberry pi for a web based home automation system. The cgi is quite handy for some tasks we didn't want to write a complete rest api for.

**Comment by edsiper** [comment]
FYI: I am porting the CGI plugin to the Monkey v1.6 architecture and API changes. Once the base functionality is in place, I will investigate and improve the problem reported here. Stay tuned.

**Comment by edsiper** [comment]
I have pushed the new changes in master. 

@ambiHomeCE @eccoecco , would you please test if the CGI port and the problem reported here is fixed ?, any input is appreciated.

note: I just did a very simple test with your script.

**Comment by eccoecco** [comment]
Hi, I'm so sorry, I am no longer part of the project that is using Monkey, and no longer have access to the development/test environment.

**Comment by edsiper** [comment]
@eccoecco would be possible you can point me to someone to ask about this ?, some colleague or manager from your ex-team ?

thanks.

**Comment by ambiHomeCE** [comment]
Seems to work, thanks :)
I found another flaw with the request being stalled and in the end dropped by monkey if the CGI doesn't return a Content-Type for the response, e.g. when using status 204 No content. Should I create another issue for this?

**Comment by edsiper** [comment]
@ambiHomeCE yes please. Please include details about how to reproduce it and the status on 1.5 and 1.6 versions.

**Comment by edsiper** [comment]
Closing this issue as seems to be fixed in Master. If something is not working properly feel free to re-open it.

**Comment by ambiHomeCE** [comment]
Sorry, but It was premature for me give all-clear. I tested again with the bash script given in the first post and it isn't working properly. I expect it monkey to return "Hallo World" in the response, but testing with wget and curl returns "0".

**Comment by edsiper** [comment]
No problem, I am re-opening the issue until we have enough confirmation, I don't want a buggy code on 1.6.

The specific breakline issue I cannot reproduce it after the patches, I am fixing other issues when handling closed connections which should not be related.

I tested with curl and wget and it worked for me, would you please provide the curl and wget versions ?, I would like to try to reproduce with your environment setup.

**Comment by ambiHomeCE** [comment]
I tried on a raspberry pi (raspbian) system. wget version is 
````
root@ambihome:~# wget --version
GNU Wget 1.13.4 built on linux-gnueabihf.

+digest +https +ipv6 +iri +large-file +nls -ntlm +opie +ssl/gnutls
````
curl
````
root@ambihome:~# curl --version
curl 7.26.0 (arm-unknown-linux-gnueabihf) libcurl/7.26.0 OpenSSL/1.0.1e zlib/1.2.7 libidn/1.25 libssh2/1.4.2 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp
Features: Debug GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP
````

**Comment by eccoecco** [comment]
Hi,

Sorry it took so long to reply.  I've managed to talk to some of the team from my previous job, and it seems that they no longer use monkey, as there was apparently an issue with a failed POST request when uploading files via CGI (or something similar).

Unfortunately, I wasn't able to convince them to test the new patch due to deadline pressures.  Sorry. =(

For reference's sake, the platform where this was originally discovered was a custom iMX6Q board with a custom Linux distribution branched from Freescale's reference Yocto 1.5 branch.

**Comment by edsiper** [comment]
Found that wget version uses only http/1.0, I was able to reproduce a problem with that old protocol, working in a fix.

**Comment by edsiper** [comment]
I have pushed some improvements for the CGI plugin and minor tweaks for the Monkey Scheduler, if you can test any feedback is appreciated.

Note: make sure you are using updated tools for test (curl, wget) as I found that older versions have problems when processing replies with chunked transfer encoding mode used in CGI. I found also that the tool weighttp is totally broken, do not use it.

**Comment by edsiper** [comment]
I did more tests and I am not longer able to reproduce the "separted \r\n issue", if you find is reproducible again, please re-open this ticket.

For other CGI problems please open a new bug (have not find any other).
